numba/cpython/charseq.py

```diff
@@ -249,6 +249,59 @@ def codegen(context, builder, signature, args):
     return sig, codegen
 
 
+@lower_cast(types.StringLiteral, types.UnicodeCharSeq)
+def cast_from_literal(context, builder, fromty, toty, val):
+    from .unicode import compile_time_get_string_data
+    literal_value = fromty.literal_value
+
+    databytes, length, kind, is_ascii, hashv = \
+        compile_time_get_string_data(literal_value)
+    mod = builder.module
+
+    i8 = ir.IntType(8)
+    i16 = ir.IntType(16)
+    i32 = ir.IntType(32)
+
+    # src
+    gv = context.insert_const_bytes(mod, databytes)
+    src1 = builder.bitcast(gv, i8.as_pointer())
+    src2 = builder.bitcast(gv, i16.as_pointer())
+    src4 = builder.bitcast(gv, i32.as_pointer())
+
+    kind1 = builder.icmp_unsigned('==', i8(kind), i8(1))
+    kind2 = builder.icmp_unsigned('==', i8(kind), i8(2))
+    kind4 = builder.icmp_unsigned('==', i8(kind), i8(4))
+
+    # size
+    lty = context.get_value_type(toty)
+    size = i8(len(literal_value))
+
+    # dest
+    dstint_t = ir.IntType(8 * unicode_byte_width)
+    dst_ptr = cgutils.alloca_once(builder, lty, size=length)
+    dst = builder.bitcast(dst_ptr, i32.as_pointer())
+
+    with builder.if_then(kind1):
+        with cgutils.for_range(builder, size) as loop:
+            in_ptr = builder.gep(src1, [loop.index])
+            in_val = builder.zext(builder.load(in_ptr), dstint_t)
+            builder.store(in_val, builder.gep(dst, [loop.index]))
+
+    with builder.if_then(kind2):
+        with cgutils.for_range(builder, size) as loop:
+            in_ptr = builder.gep(src2, [loop.index])
+            in_val = builder.zext(builder.load(in_ptr), dstint_t)
+            builder.store(in_val, builder.gep(dst, [loop.index]))
+
+    with builder.if_then(kind4):
+        with cgutils.for_range(builder, size) as loop:
+            in_ptr = builder.gep(src4, [loop.index])
+            in_val = builder.zext(builder.load(in_ptr), dstint_t)
+            builder.store(in_val, builder.gep(dst, [loop.index]))
+
```

Comment:
What do you think of perhaps folding all this into something like:
```python
    bits = 8
    i8 = ir.IntType(bits)
    i32 = ir.IntType(32)

    # sizes
    lty = context.get_value_type(toty)
    size = i8(len(literal_value))

    # returned string
    dstint_t = ir.IntType(8 * unicode_byte_width)
    dst_ptr = cgutils.alloca_once(builder, lty, size=length)
    dst = builder.bitcast(dst_ptr, i32.as_pointer())

    # const
    gv = context.insert_const_bytes(mod, databytes)

    for p in range(3):
        bw = 1 << p
        ibw = ir.IntType(bits * bw)
        src = builder.bitcast(gv, ibw.as_pointer())
        switching_kind = builder.icmp_unsigned('==', i8(kind), i8(bw))

        with builder.if_then(switching_kind):
            with cgutils.for_range(builder, size) as loop:
                in_ptr = builder.gep(src, [loop.index])
                in_val = builder.zext(builder.load(in_ptr), dstint_t)
                builder.store(in_val, builder.gep(dst, [loop.index]))
```
?

Reply:
Also wonder if there's benefit in just doing:
```python
@overload(np.asarray)
def ol_asarray(string):
    if not isinstance(string, types.StringLiteral):
        return None
    arr = np.array(string.literal_value)
    def impl(string):
        return arr.copy()
    return impl
```
or similar?

Reply:
That's way better. Less code to maintain. I will replace my current implementation by this one.